File: Android/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/alignruler/AlignRulerMarkerDokitView.java
package com.didichuxing.doraemonkit.kit.alignruler;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.WindowManager;
import android.widget.FrameLayout;

import com.didichuxing.doraemonkit.DoraemonKit;
import com.didichuxing.doraemonkit.R;
import com.didichuxing.doraemonkit.ui.base.AbsDokitView;
import com.didichuxing.doraemonkit.ui.base.DokitViewLayoutParams;
import com.didichuxing.doraemonkit.util.UIUtils;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by jintai on 2019/09/26.
 */

public class AlignRulerMarkerDokitView extends AbsDokitView {
    private List<OnAlignRulerMarkerPositionChangeListener> mPositionChangeListeners = new ArrayList<>();


    @Override
    public View onCreateView(Context context, FrameLayout view) {
        return LayoutInflater.from(context).inflate(R.layout.dk_float_align_ruler_marker, null);
    }

    @Override
    public void onViewCreated(FrameLayout view) {

    }


    @Override
    public void initDokitViewLayoutParams(DokitViewLayoutParams params) {
        params.height = DokitViewLayoutParams.WRAP_CONTENT;
        params.width = DokitViewLayoutParams.WRAP_CONTENT;
        params.x = UIUtils.getWidthPixels() / 2;
        params.y = UIUtils.getHeightPixels() / 2;
    }

    @Override
    public void onCreate(Context context) {
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        removePositionChangeListeners();
    }

    @Override
    public void onMove(int x, int y, int dx, int dy) {
        super.onMove(x, y, dx, dy);
        for (OnAlignRulerMarkerPositionChangeListener listener : mPositionChangeListeners) {
            if (isNormalMode()) {
                listener.onPositionChanged(getNormalLayoutParams().leftMargin + getRootView().getWidth() / 2, getNormalLayoutParams().topMargin + getRootView().getHeight() / 2);
            } else {
                listener.onPositionChanged(getSystemLayoutParams().x + getRootView().getWidth() / 2, getSystemLayoutParams().y + getRootView().getHeight() / 2);
            }
        }
    }

    @Override
    public void updateViewLayout(String tag, boolean isActivityResume) {
        super.updateViewLayout(tag, isActivityResume);
        //更新标尺的位置信息
        for (OnAlignRulerMarkerPositionChangeListener listener : mPositionChangeListeners) {
            if (isNormalMode()) {
                listener.onPositionChanged(getNormalLayoutParams().leftMargin + getRootView().getWidth() / 2, getNormalLayoutParams().topMargin + getRootView().getHeight() / 2);
            } else {
                listener.onPositionChanged(getSystemLayoutParams().x + getRootView().getWidth() / 2, getSystemLayoutParams().y + getRootView().getHeight() / 2);
            }
        }

    }

    public interface OnAlignRulerMarkerPositionChangeListener {
        void onPositionChanged(int x, int y);
    }

    public void addPositionChangeListener(OnAlignRulerMarkerPositionChangeListener positionChangeListener) {
        mPositionChangeListeners.add(positionChangeListener);
        //更新标尺的位置信息
        for (OnAlignRulerMarkerPositionChangeListener listener : mPositionChangeListeners) {
            if (isNormalMode()) {
                listener.onPositionChanged(getNormalLayoutParams().leftMargin + getRootView().getWidth() / 2, getNormalLayoutParams().topMargin + getRootView().getHeight() / 2);
            } else {
                listener.onPositionChanged(getSystemLayoutParams().x + getRootView().getWidth() / 2, getSystemLayoutParams().y + getRootView().getHeight() / 2);
            }
        }
    }

    public void removePositionChangeListener(OnAlignRulerMarkerPositionChangeListener positionChangeListener) {
        mPositionChangeListeners.remove(positionChangeListener);
    }

    private void removePositionChangeListeners() {
        mPositionChangeListeners.clear();
    }

    @Override
    public boolean restrictBorderline() {
        return false;
    }
}

File: Android/kotlin/doraemonkit/src/main/java/com/didichuxing/doraemonkit/kit/alignruler/AlignRulerMarkerDokitView.kt
package com.didichuxing.doraemonkit.kit.alignruler

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.widget.FrameLayout
import com.didichuxing.doraemonkit.R
import com.didichuxing.doraemonkit.kit.core.AbsDokitView
import com.didichuxing.doraemonkit.kit.core.DokitViewLayoutParams
import com.didichuxing.doraemonkit.util.UIUtils.heightPixels
import com.didichuxing.doraemonkit.util.UIUtils.widthPixels
import java.util.*

/**
 * @author: xuchun
 * @time: 2020/6/4 - 10:47
 * @desc: 对齐标尺标记View
 */
class AlignRulerMarkerDokitView : AbsDokitView() {
    private val mPositionChangeListeners: MutableList<OnAlignRulerMarkerPositionChangeListener> = ArrayList()
    override fun onCreateView(context: Context, rootView: FrameLayout): View {
        return LayoutInflater.from(context).inflate(R.layout.dk_float_align_ruler_marker, null)
    }

    override fun onViewCreated(rootView: FrameLayout) {}
    override fun initDokitViewLayoutParams(params: DokitViewLayoutParams?) {
        params?.let {
            it.height = DokitViewLayoutParams.WRAP_CONTENT
            it.width = DokitViewLayoutParams.WRAP_CONTENT
            it.x = widthPixels / 2
            it.y = heightPixels / 2
        }

    }

    override fun onCreate(context: Context) {}
    override fun onDestroy() {
        super.onDestroy()
        removePositionChangeListeners()
    }

    override fun onMove(x: Int, y: Int, dx: Int, dy: Int) {
        super.onMove(x, y, dx, dy)
        for (listener in mPositionChangeListeners) {
            if (isNormalMode) {
                listener.onPositionChanged(normalLayoutParams.leftMargin + rootView.width / 2, normalLayoutParams.topMargin + rootView.height / 2)
            } else {
                listener.onPositionChanged(systemLayoutParams.x + rootView.width / 2, systemLayoutParams.y + rootView.height / 2)
            }
        }
    }

    override fun updateViewLayout(tag: String, isActivityResume: Boolean) {
        super.updateViewLayout(tag, isActivityResume)
        //更新标尺的位置信息
        for (listener in mPositionChangeListeners) {
            if (isNormalMode) {
                listener.onPositionChanged(normalLayoutParams.leftMargin + rootView.width / 2, normalLayoutParams.topMargin + rootView.height / 2)
            } else {
                listener.onPositionChanged(systemLayoutParams.x + rootView.width / 2, systemLayoutParams.y + rootView.height / 2)
            }
        }
    }

    interface OnAlignRulerMarkerPositionChangeListener {
        fun onPositionChanged(x: Int, y: Int)
    }

    fun addPositionChangeListener(positionChangeListener: OnAlignRulerMarkerPositionChangeListener) {
        mPositionChangeListeners.add(positionChangeListener)
        //更新标尺的位置信息
        for (listener in mPositionChangeListeners) {
            if (isNormalMode) {
                listener.onPositionChanged(normalLayoutParams.leftMargin + rootView.width / 2, normalLayoutParams.topMargin + rootView.height / 2)
            } else {
                listener.onPositionChanged(systemLayoutParams.x + rootView.width / 2, systemLayoutParams.y + rootView.height / 2)
            }
        }
    }

    fun removePositionChangeListener(positionChangeListener: OnAlignRulerMarkerPositionChangeListener) {
        mPositionChangeListeners.remove(positionChangeListener)
    }

    private fun removePositionChangeListeners() {
        mPositionChangeListeners.clear()
    }

    override fun restrictBorderline(): Boolean {
        return false
    }
}